File: sdk/typescript/src/namespaces/cost-management.ts
/**
 * Cost Management Namespace API
 *
 * Provides methods for tracking and managing AI costs:
 * - Cost dashboard and summaries
 * - Cost breakdown by provider, feature, model
 * - Budget management and alerts
 * - Usage timeline and forecasting
 * - Optimization recommendations
 * - Pre-execution cost estimation
 * - What-if scenario simulation
 */

// =============================================================================
// Enumerations
// =============================================================================

/**
 * Cost data granularity
 */
export type CostGranularity = 'hourly' | 'daily' | 'weekly' | 'monthly';

/**
 * Cost trend direction
 */
export type TrendDirection = 'increasing' | 'decreasing' | 'stable';

/**
 * Seasonal cost pattern
 */
export type SeasonalPattern = 'daily' | 'weekly' | 'monthly' | 'none';

/**
 * Alert severity levels
 */
export type AlertSeverity = 'info' | 'warning' | 'critical';

/**
 * Budget alert levels
 */
export type BudgetAlertLevel = 'info' | 'warning' | 'critical' | 'exceeded';

/**
 * Recommendation types
 */
export type RecommendationType =
  | 'model_downgrade'
  | 'caching'
  | 'batching'
  | 'rate_limiting'
  | 'prompt_optimization'
  | 'provider_switch'
  | 'time_shifting'
  | 'quota_adjustment';

/**
 * Recommendation priority levels
 */
export type RecommendationPriority = 'critical' | 'high' | 'medium' | 'low';

/**
 * Recommendation status
 */
export type RecommendationStatus = 'pending' | 'applied' | 'dismissed' | 'expired' | 'partial';

/**
 * Cost enforcement modes
 */
export type CostEnforcementMode = 'hard' | 'soft' | 'throttle' | 'estimate';

/**
 * Throttle levels based on budget usage
 */
export type ThrottleLevel = 'none' | 'light' | 'medium' | 'heavy' | 'blocked';

// =============================================================================
// Core Types
// =============================================================================

/**
 * Cost breakdown item
 */
export interface CostBreakdownItem {
  name: string;
  cost: number;
  percentage: number;
}

/**
 * Daily cost data point
 */
export interface DailyCost {
  date: string;
  cost: number;
  tokens?: number;
}

/**
 * Cost alert
 */
export interface CostAlert {
  id: string;
  type: 'budget_warning' | 'spike_detected' | 'limit_reached';
  message: string;
  severity: 'critical' | 'warning' | 'info';
  timestamp: string;
  acknowledged?: boolean;
}

/**
 * Cost summary data
 */
export interface CostSummary {
  total_cost: number;
  budget: number;
  tokens_used: number;
  api_calls: number;
  last_updated: string;
  cost_by_provider?: CostBreakdownItem[];
  cost_by_feature?: CostBreakdownItem[];
  daily_costs?: DailyCost[];
  alerts?: CostAlert[];
}

/**
 * Budget configuration
 */
export interface CostBudget {
  workspace_id: string;
  monthly_limit: number;
  daily_limit?: number;
  current_spend?: number;
  alert_thresholds?: number[];
}

/**
 * Cost optimization recommendation
 */
export interface CostRecommendation {
  id: string;
  type: 'model_downgrade' | 'caching' | 'batching' | string;
  title: string;
  description: string;
  estimated_savings: number;
  effort: 'low' | 'medium' | 'high';
  status: 'pending' | 'applied' | 'dismissed';
  created_at?: string;
}

/**
 * Efficiency metrics
 */
export interface EfficiencyMetrics {
  cost_per_1k_tokens: number;
  tokens_per_call: number;
  cost_per_call: number;
  total_tokens: number;
  total_calls: number;
  total_cost: number;
  model_utilization?: Array<{ model: string; calls: number; cost: number }>;
}

/**
 * Basic cost forecast data
 */
export interface CostForecast {
  workspace_id: string;
  forecast_days: number;
  projected_cost: number;
  confidence_interval?: [number, number];
  trend: TrendDirection;
  daily_projections?: Array<{ date: string; cost: number }>;
}

// =============================================================================
// Advanced Forecasting Types
// =============================================================================

/** Detailed trend analysis for costs */
export interface TrendAnalysis {
  direction: TrendDirection;
  change_rate: number;
  change_rate_weekly: number;
  r_squared: number;
  trend_start?: string;
  confidence: number;
}

/** Daily forecast with confidence intervals */
export interface DailyForecast {
  date: string;
  predicted_cost: number;
  lower_bound: number;
  upper_bound: number;
  confidence: number;
}

/** Forecast alert for projected issues */
export interface ForecastAlert {
  type: 'budget_exceeded' | 'anomaly_detected' | 'trend_warning';
  message: string;
  severity: AlertSeverity;
  date: string;
  value?: number;
}

/** Comprehensive forecast report with trend analysis */
export interface ForecastReport {
  predicted_monthly_cost: number;
  predicted_daily_average: number;
  confidence_interval: number;
  trend?: TrendAnalysis;
  seasonal_pattern: SeasonalPattern;
  daily_forecasts: DailyForecast[];
  alerts: ForecastAlert[];
  projected_budget_usage?: number;
  days_until_budget_exceeded?: number | null;
  generated_at: string;
}

// =============================================================================
// Cost Estimation Types
// =============================================================================

/** Pre-execution cost estimate for a task */
export interface CostEstimate {
  estimated_cost_usd: number;
  estimated_input_tokens: number;
  estimated_output_tokens: number;
  estimated_tokens: number;
  confidence: number;
  based_on_samples: number;
  model_suggestion?: string;
  estimated_savings_usd?: number;
  model_breakdown?: Array<{
    model: string;
    estimated_cost: number;
    estimated_tokens: number;
  }>;
}

/** Cost constraint validation result */
export interface CostConstraintResult {
  allowed: boolean;
  reason?: string;
  throttle_level: ThrottleLevel;
  enforcement_mode: CostEnforcementMode;
  budget_percentage_used: number;
  remaining_budget_usd?: number;
  estimated_cost?: CostEstimate;
  priority_adjustment: number;
}

// =============================================================================
// Simulation Types
// =============================================================================

/** Cost simulation scenario definition */
export interface SimulationScenario {
  name: string;
  description?: string;
  changes: {
    model_change?: string;
    volume_multiplier?: number;
    caching_enabled?: boolean;
    batch_size?: number;
    provider_change?: string;
    [key: string]: unknown;
  };
}

/** Cost simulation result */
export interface SimulationResult {
  baseline_cost: number;
  simulated_cost: number;
  cost_difference: number;
  percentage_change: number;
  daily_breakdown: Array<{
    date: string;
    baseline: number;
    simulated: number;
    difference: number;
  }>;
  quality_impact?: string;
  risk_level?: 'low' | 'medium' | 'high';
}

// =============================================================================
// Advanced Recommendation Types
// =============================================================================

/** Alternative model suggestion for cost optimization */
export interface ModelAlternative {
  provider: string;
  model: string;
  cost_per_1k_input: number;
  cost_per_1k_output: number;
  quality_score: number;
  latency_multiplier: number;
  suitable_for: string[];
}

/** Caching opportunity analysis */
export interface CachingOpportunity {
  pattern: 'system_prompt' | 'repeated_query' | 'prefix' | 'semantic';
  estimated_hit_rate: number;
  unique_queries: number;
  repeat_count: number;
  cache_strategy: 'exact' | 'semantic' | 'prefix';
  estimated_savings_usd: number;
}

/** Batching opportunity analysis */
export interface BatchingOpportunity {
  current_rpm: number;
  recommended_batch_size: number;
  latency_increase_ms: number;
  savings_percentage: number;
  batchable_tasks: string[];
}

/** Implementation step for a recommendation */
export interface ImplementationStep {
  order: number;
  description: string;
  code_snippet?: string;
  config_change?: Record<string, unknown>;
  effort_minutes?: number;
}

/** Enhanced cost optimization recommendation with full details */
export interface DetailedCostRecommendation {
  id: string;
  type: RecommendationType;
  priority: RecommendationPriority;
  title: string;
  description: string;
  estimated_savings: number;
  effort: 'low' | 'medium' | 'high';
  status: RecommendationStatus;
  model_alternative?: ModelAlternative;
  caching_opportunity?: CachingOpportunity;
  batching_opportunity?: BatchingOpportunity;
  implementation_steps?: ImplementationStep[];
  quality_impact?: string;
  quality_impact_score?: number;
  risk_level: 'low' | 'medium' | 'high';
  auto_apply_available: boolean;
  requires_approval: boolean;
  created_at: string;
  expires_at?: string;
}

/** Summary of all recommendations */
export interface RecommendationSummary {
  total_count: number;
  total_estimated_savings: number;
  by_type: Record<string, number>;
  by_priority: Record<string, number>;
  auto_apply_count: number;
}

/**
 * Time range options for cost queries
 */
export type TimeRange = '24h' | '7d' | '30d' | '90d';

/**
 * Group by options for cost breakdown
 */
export type GroupBy = 'provider' | 'feature' | 'model';

/**
 * Interface for the internal client used by CostManagementAPI.
 */
interface CostManagementClientInterface {
  request<T>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Cost Management API namespace.
 *
 * Provides methods for tracking and managing AI costs:
 * - View cost summaries and breakdowns
 * - Set and manage budgets
 * - Get cost optimization recommendations
 * - Forecast future costs
 * - Estimate costs before execution
 * - Simulate what-if scenarios
 *
 * Essential for SME cost management and preventing unexpected charges.
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Get cost summary
 * const summary = await client.costManagement.getSummary({ range: '30d' });
 * console.log(`Total cost: $${summary.total_cost}`);
 *
 * // Get breakdown by provider
 * const { breakdown, total } = await client.costManagement.getBreakdown({ group_by: 'provider' });
 *
 * // Set a budget
 * await client.costManagement.setBudget({ budget: 500, daily_limit: 20 });
 *
 * // Get optimization recommendations
 * const { recommendations } = await client.costManagement.getRecommendations();
 *
 * // Estimate cost before running a task
 * const estimate = await client.costManagement.estimateCost({ task: 'analyze document', model: 'claude-3-opus' });
 *
 * // Simulate switching to a cheaper model
 * const simulation = await client.costManagement.simulateScenario({
 *   workspace_id: 'ws-123',
 *   scenario: { name: 'Switch to Haiku', changes: { model_change: 'claude-3-haiku' } }
 * });
 * ```
 */
export class CostManagementAPI {
  constructor(private client: CostManagementClientInterface) {}

  // ===========================================================================
  // Cost Dashboard
  // ===========================================================================

  /**
   * Get cost dashboard summary.
   */
  async getSummary(options?: {
    workspace_id?: string;
    range?: TimeRange;
  }): Promise<CostSummary> {
    return this.client.request('GET', '/api/costs', { params: options });
  }

  /**
   * Get detailed cost breakdown.
   */
  async getBreakdown(options?: {
    workspace_id?: string;
    range?: TimeRange;
    group_by?: GroupBy;
  }): Promise<{ breakdown: CostBreakdownItem[]; total: number }> {
    return this.client.request('GET', '/api/costs/breakdown', { params: options });
  }

  /**
   * Get usage timeline data.
   */
  async getTimeline(options?: {
    workspace_id?: string;
    range?: TimeRange;
  }): Promise<{ timeline: DailyCost[]; total: number; average: number }> {
    return this.client.request('GET', '/api/costs/timeline', { params: options });
  }

  // ===========================================================================
  // Alerts
  // ===========================================================================

  /**
   * Get budget alerts.
   */
  async getAlerts(options?: {
    workspace_id?: string;
  }): Promise<{ alerts: CostAlert[] }> {
    return this.client.request('GET', '/api/costs/alerts', { params: options });
  }

  /**
   * Dismiss a budget alert.
   */
  async dismissAlert(
    alertId: string,
    options?: { workspace_id?: string }
  ): Promise<{ success: boolean }> {
    return this.client.request('POST', `/api/costs/alerts/${alertId}/dismiss`, {
      json: options,
    });
  }

  // ===========================================================================
  // Budget Management
  // ===========================================================================

  /**
   * Set budget limits.
   */
  async setBudget(request: {
    budget: number;
    workspace_id?: string;
    daily_limit?: number;
    name?: string;
  }): Promise<CostBudget> {
    return this.client.request('POST', '/api/costs/budget', { json: request });
  }

  // ===========================================================================
  // Recommendations
  // ===========================================================================

  /**
   * Get cost optimization recommendations.
   */
  async getRecommendations(options?: {
    workspace_id?: string;
    status?: 'pending' | 'applied' | 'dismissed';
    type?: string;
  }): Promise<{ recommendations: CostRecommendation[] }> {
    return this.client.request('GET', '/api/costs/recommendations', { params: options });
  }

  /**
   * Get detailed recommendations with full analysis.
   */
  async getDetailedRecommendations(options?: {
    workspace_id?: string;
    status?: RecommendationStatus;
    type?: RecommendationType;
    priority?: RecommendationPriority;
  }): Promise<{ recommendations: DetailedCostRecommendation[]; summary: RecommendationSummary }> {
    return this.client.request('GET', '/api/costs/recommendations/detailed', { params: options });
  }

  /**
   * Get a specific recommendation by ID.
   */
  async getRecommendation(recommendationId: string): Promise<DetailedCostRecommendation> {
    return this.client.request('GET', `/api/costs/recommendations/${recommendationId}`);
  }

  /**
   * Apply a recommendation.
   */
  async applyRecommendation(
    recommendationId: string,
    options?: { user_id?: string }
  ): Promise<{ recommendation: CostRecommendation }> {
    return this.client.request('POST', `/api/costs/recommendations/${recommendationId}/apply`, {
      json: options,
    });
  }

  /**
   * Dismiss a recommendation.
   */
  async dismissRecommendation(recommendationId: string): Promise<{ success: boolean }> {
    return this.client.request('POST', `/api/costs/recommendations/${recommendationId}/dismiss`);
  }

  // ===========================================================================
  // Efficiency Metrics
  // ===========================================================================

  /**
   * Get efficiency metrics.
   */
  async getEfficiency(options?: {
    workspace_id?: string;
    range?: TimeRange;
  }): Promise<EfficiencyMetrics> {
    return this.client.request('GET', '/api/costs/efficiency', { params: options });
  }

  // ===========================================================================
  // Cost Estimation
  // ===========================================================================

  /**
   * Estimate cost for a task before execution.
   */
  async estimateCost(request: {
    task: string;
    model?: string;
    agents?: string[];
    rounds?: number;
    workspace_id?: string;
  }): Promise<CostEstimate> {
    return this.client.request('POST', '/api/costs/estimate', { json: request });
  }

  /**
   * Check if a task is allowed under current budget constraints.
   */
  async checkConstraints(request: {
    task: string;
    model?: string;
    workspace_id?: string;
  }): Promise<CostConstraintResult> {
    return this.client.request('POST', '/api/costs/constraints/check', { json: request });
  }

  // ===========================================================================
  // Forecasting
  // ===========================================================================

  /**
   * Get basic cost forecast.
   */
  async getForecast(options?: {
    workspace_id?: string;
    days?: number;
  }): Promise<CostForecast> {
    return this.client.request('GET', '/api/costs/forecast', { params: options });
  }

  /**
   * Get detailed forecast report with trend analysis.
   */
  async getDetailedForecast(options?: {
    workspace_id?: string;
    days?: number;
    include_alerts?: boolean;
  }): Promise<ForecastReport> {
    return this.client.request('GET', '/api/costs/forecast/detailed', { params: options });
  }

  /**
   * Simulate a cost scenario.
   */
  async simulateScenario(request: {
    workspace_id: string;
    scenario: SimulationScenario;
    days?: number;
  }): Promise<SimulationResult> {
    return this.client.request('POST', '/api/costs/forecast/simulate', { json: request });
  }
}

File: sdk/python/aragora/namespaces/cost_management.py
"""
Cost Management Namespace API.

Provides methods for tracking and managing AI costs:
- Cost dashboard and summaries
- Cost breakdown by provider, feature, model
- Budget management and alerts
- Usage timeline and forecasting
- Optimization recommendations
- Pre-execution cost estimation
- What-if scenario simulation
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

TimeRange = Literal["24h", "7d", "30d", "90d"]
GroupBy = Literal["provider", "feature", "model"]
TrendDirection = Literal["increasing", "decreasing", "stable"]
AlertSeverity = Literal["info", "warning", "critical"]
RecommendationStatus = Literal["pending", "applied", "dismissed", "expired", "partial"]
RecommendationPriority = Literal["critical", "high", "medium", "low"]
RecommendationType = Literal[
    "model_downgrade",
    "caching",
    "batching",
    "rate_limiting",
    "prompt_optimization",
    "provider_switch",
    "time_shifting",
    "quota_adjustment",
]
ThrottleLevel = Literal["none", "light", "medium", "heavy", "blocked"]
CostEnforcementMode = Literal["hard", "soft", "throttle", "estimate"]


class CostManagementAPI:
    """
    Synchronous Cost Management API.

    Provides methods for:
    - Cost dashboard and summaries
    - Budget management and alerts
    - Optimization recommendations
    - Cost forecasting and simulation
    - Pre-execution cost estimation
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # =========================================================================
    # Cost Dashboard
    # =========================================================================

    def get_summary(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
    ) -> dict[str, Any]:
        """
        Get cost dashboard summary.

        Args:
            workspace_id: Optional workspace filter.
            range: Time range (24h, 7d, 30d, 90d).

        Returns:
            Cost summary with breakdowns and alerts.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range

        return self._client._request("GET", "/api/costs", params=params)

    def get_breakdown(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
        group_by: GroupBy | None = None,
    ) -> dict[str, Any]:
        """
        Get detailed cost breakdown.

        Args:
            workspace_id: Optional workspace filter.
            range: Time range.
            group_by: Group by dimension (provider, feature, model).

        Returns:
            Cost breakdown with total.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range
        if group_by:
            params["group_by"] = group_by

        return self._client._request("GET", "/api/costs/breakdown", params=params)

    def get_timeline(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
    ) -> dict[str, Any]:
        """
        Get usage timeline data.

        Args:
            workspace_id: Optional workspace filter.
            range: Time range.

        Returns:
            Timeline with daily costs and averages.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range

        return self._client._request("GET", "/api/costs/timeline", params=params)

    # =========================================================================
    # Alerts
    # =========================================================================

    def get_alerts(
        self,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Get budget alerts.

        Args:
            workspace_id: Optional workspace filter.

        Returns:
            List of cost alerts.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id

        return self._client._request("GET", "/api/costs/alerts", params=params)

    def dismiss_alert(
        self,
        alert_id: str,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Dismiss a budget alert.

        Args:
            alert_id: Alert identifier.
            workspace_id: Optional workspace.

        Returns:
            Success status.
        """
        data: dict[str, Any] = {}
        if workspace_id:
            data["workspace_id"] = workspace_id

        return self._client._request("POST", f"/api/costs/alerts/{alert_id}/dismiss", json=data)

    # =========================================================================
    # Budget Management
    # =========================================================================

    def set_budget(
        self,
        budget: float,
        workspace_id: str | None = None,
        daily_limit: float | None = None,
        name: str | None = None,
    ) -> dict[str, Any]:
        """
        Set budget limits.

        Args:
            budget: Monthly budget limit.
            workspace_id: Optional workspace.
            daily_limit: Optional daily limit.
            name: Budget name.

        Returns:
            Created budget configuration.
        """
        data: dict[str, Any] = {"budget": budget}
        if workspace_id:
            data["workspace_id"] = workspace_id
        if daily_limit is not None:
            data["daily_limit"] = daily_limit
        if name:
            data["name"] = name

        return self._client._request("POST", "/api/costs/budget", json=data)

    # =========================================================================
    # Recommendations
    # =========================================================================

    def get_recommendations(
        self,
        workspace_id: str | None = None,
        status: RecommendationStatus | None = None,
        type: str | None = None,
    ) -> dict[str, Any]:
        """
        Get cost optimization recommendations.

        Args:
            workspace_id: Optional workspace filter.
            status: Filter by status.
            type: Filter by recommendation type.

        Returns:
            List of recommendations.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if status:
            params["status"] = status
        if type:
            params["type"] = type

        return self._client._request("GET", "/api/costs/recommendations", params=params)

    def get_detailed_recommendations(
        self,
        workspace_id: str | None = None,
        status: RecommendationStatus | None = None,
        type: RecommendationType | None = None,
        priority: RecommendationPriority | None = None,
    ) -> dict[str, Any]:
        """
        Get detailed recommendations with full analysis.

        Args:
            workspace_id: Optional workspace filter.
            status: Filter by status.
            type: Filter by type.
            priority: Filter by priority.

        Returns:
            Detailed recommendations with summary.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if status:
            params["status"] = status
        if type:
            params["type"] = type
        if priority:
            params["priority"] = priority

        return self._client._request("GET", "/api/costs/recommendations/detailed", params=params)

    def get_recommendation(self, recommendation_id: str) -> dict[str, Any]:
        """
        Get a specific recommendation by ID.

        Args:
            recommendation_id: Recommendation identifier.

        Returns:
            Detailed recommendation.
        """
        return self._client._request("GET", f"/api/costs/recommendations/{recommendation_id}")

    def apply_recommendation(
        self,
        recommendation_id: str,
        user_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Apply a recommendation.

        Args:
            recommendation_id: Recommendation identifier.
            user_id: Optional user applying the recommendation.

        Returns:
            Applied recommendation.
        """
        data: dict[str, Any] = {}
        if user_id:
            data["user_id"] = user_id

        return self._client._request(
            "POST", f"/api/costs/recommendations/{recommendation_id}/apply", json=data
        )

    def dismiss_recommendation(self, recommendation_id: str) -> dict[str, Any]:
        """
        Dismiss a recommendation.

        Args:
            recommendation_id: Recommendation identifier.

        Returns:
            Success status.
        """
        return self._client._request(
            "POST", f"/api/costs/recommendations/{recommendation_id}/dismiss"
        )

    # =========================================================================
    # Efficiency Metrics
    # =========================================================================

    def get_efficiency(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
    ) -> dict[str, Any]:
        """
        Get efficiency metrics.

        Args:
            workspace_id: Optional workspace filter.
            range: Time range.

        Returns:
            Efficiency metrics including cost per token.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range

        return self._client._request("GET", "/api/costs/efficiency", params=params)

    # =========================================================================
    # Cost Estimation
    # =========================================================================

    def estimate_cost(
        self,
        task: str,
        model: str | None = None,
        agents: list[str] | None = None,
        rounds: int | None = None,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Estimate cost for a task before execution.

        Args:
            task: Task description.
            model: Target model.
            agents: List of agents to use.
            rounds: Number of debate rounds.
            workspace_id: Optional workspace.

        Returns:
            Cost estimate with confidence.
        """
        data: dict[str, Any] = {"task": task}
        if model:
            data["model"] = model
        if agents:
            data["agents"] = agents
        if rounds is not None:
            data["rounds"] = rounds
        if workspace_id:
            data["workspace_id"] = workspace_id

        return self._client._request("POST", "/api/costs/estimate", json=data)

    def check_constraints(
        self,
        task: str,
        model: str | None = None,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Check if a task is allowed under current budget constraints.

        Args:
            task: Task description.
            model: Target model.
            workspace_id: Optional workspace.

        Returns:
            Constraint check result with throttle level.
        """
        data: dict[str, Any] = {"task": task}
        if model:
            data["model"] = model
        if workspace_id:
            data["workspace_id"] = workspace_id

        return self._client._request("POST", "/api/costs/constraints/check", json=data)

    # =========================================================================
    # Forecasting
    # =========================================================================

    def get_forecast(
        self,
        workspace_id: str | None = None,
        days: int | None = None,
    ) -> dict[str, Any]:
        """
        Get basic cost forecast.

        Args:
            workspace_id: Optional workspace filter.
            days: Forecast period in days.

        Returns:
            Cost forecast with projections.
        """
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if days is not None:
            params["days"] = days

        return self._client._request("GET", "/api/costs/forecast", params=params)

    def get_detailed_forecast(
        self,
        workspace_id: str | None = None,
        days: int | None = None,
        include_alerts: bool = True,
    ) -> dict[str, Any]:
        """
        Get detailed forecast report with trend analysis.

        Args:
            workspace_id: Optional workspace filter.
            days: Forecast period in days.
            include_alerts: Include forecast alerts.

        Returns:
            Detailed forecast report.
        """
        params: dict[str, Any] = {"include_alerts": include_alerts}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if days is not None:
            params["days"] = days

        return self._client._request("GET", "/api/costs/forecast/detailed", params=params)

    def simulate_scenario(
        self,
        workspace_id: str,
        scenario: dict[str, Any],
        days: int | None = None,
    ) -> dict[str, Any]:
        """
        Simulate a cost scenario.

        Args:
            workspace_id: Workspace identifier.
            scenario: Scenario definition with name and changes.
            days: Simulation period in days.

        Returns:
            Simulation result with cost comparison.
        """
        data: dict[str, Any] = {
            "workspace_id": workspace_id,
            "scenario": scenario,
        }
        if days is not None:
            data["days"] = days

        return self._client._request("POST", "/api/costs/forecast/simulate", json=data)


class AsyncCostManagementAPI:
    """Asynchronous Cost Management API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # =========================================================================
    # Cost Dashboard
    # =========================================================================

    async def get_summary(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
    ) -> dict[str, Any]:
        """Get cost dashboard summary."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range

        return await self._client._request("GET", "/api/costs", params=params)

    async def get_breakdown(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
        group_by: GroupBy | None = None,
    ) -> dict[str, Any]:
        """Get detailed cost breakdown."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range
        if group_by:
            params["group_by"] = group_by

        return await self._client._request("GET", "/api/costs/breakdown", params=params)

    async def get_timeline(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
    ) -> dict[str, Any]:
        """Get usage timeline data."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range

        return await self._client._request("GET", "/api/costs/timeline", params=params)

    # =========================================================================
    # Alerts
    # =========================================================================

    async def get_alerts(
        self,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """Get budget alerts."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id

        return await self._client._request("GET", "/api/costs/alerts", params=params)

    async def dismiss_alert(
        self,
        alert_id: str,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """Dismiss a budget alert."""
        data: dict[str, Any] = {}
        if workspace_id:
            data["workspace_id"] = workspace_id

        return await self._client._request(
            "POST", f"/api/costs/alerts/{alert_id}/dismiss", json=data
        )

    # =========================================================================
    # Budget Management
    # =========================================================================

    async def set_budget(
        self,
        budget: float,
        workspace_id: str | None = None,
        daily_limit: float | None = None,
        name: str | None = None,
    ) -> dict[str, Any]:
        """Set budget limits."""
        data: dict[str, Any] = {"budget": budget}
        if workspace_id:
            data["workspace_id"] = workspace_id
        if daily_limit is not None:
            data["daily_limit"] = daily_limit
        if name:
            data["name"] = name

        return await self._client._request("POST", "/api/costs/budget", json=data)

    # =========================================================================
    # Recommendations
    # =========================================================================

    async def get_recommendations(
        self,
        workspace_id: str | None = None,
        status: RecommendationStatus | None = None,
        type: str | None = None,
    ) -> dict[str, Any]:
        """Get cost optimization recommendations."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if status:
            params["status"] = status
        if type:
            params["type"] = type

        return await self._client._request("GET", "/api/costs/recommendations", params=params)

    async def get_detailed_recommendations(
        self,
        workspace_id: str | None = None,
        status: RecommendationStatus | None = None,
        type: RecommendationType | None = None,
        priority: RecommendationPriority | None = None,
    ) -> dict[str, Any]:
        """Get detailed recommendations with full analysis."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if status:
            params["status"] = status
        if type:
            params["type"] = type
        if priority:
            params["priority"] = priority

        return await self._client._request(
            "GET", "/api/costs/recommendations/detailed", params=params
        )

    async def get_recommendation(self, recommendation_id: str) -> dict[str, Any]:
        """Get a specific recommendation by ID."""
        return await self._client._request("GET", f"/api/costs/recommendations/{recommendation_id}")

    async def apply_recommendation(
        self,
        recommendation_id: str,
        user_id: str | None = None,
    ) -> dict[str, Any]:
        """Apply a recommendation."""
        data: dict[str, Any] = {}
        if user_id:
            data["user_id"] = user_id

        return await self._client._request(
            "POST", f"/api/costs/recommendations/{recommendation_id}/apply", json=data
        )

    async def dismiss_recommendation(self, recommendation_id: str) -> dict[str, Any]:
        """Dismiss a recommendation."""
        return await self._client._request(
            "POST", f"/api/costs/recommendations/{recommendation_id}/dismiss"
        )

    # =========================================================================
    # Efficiency Metrics
    # =========================================================================

    async def get_efficiency(
        self,
        workspace_id: str | None = None,
        range: TimeRange | None = None,
    ) -> dict[str, Any]:
        """Get efficiency metrics."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if range:
            params["range"] = range

        return await self._client._request("GET", "/api/costs/efficiency", params=params)

    # =========================================================================
    # Cost Estimation
    # =========================================================================

    async def estimate_cost(
        self,
        task: str,
        model: str | None = None,
        agents: list[str] | None = None,
        rounds: int | None = None,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """Estimate cost for a task before execution."""
        data: dict[str, Any] = {"task": task}
        if model:
            data["model"] = model
        if agents:
            data["agents"] = agents
        if rounds is not None:
            data["rounds"] = rounds
        if workspace_id:
            data["workspace_id"] = workspace_id

        return await self._client._request("POST", "/api/costs/estimate", json=data)

    async def check_constraints(
        self,
        task: str,
        model: str | None = None,
        workspace_id: str | None = None,
    ) -> dict[str, Any]:
        """Check if a task is allowed under current budget constraints."""
        data: dict[str, Any] = {"task": task}
        if model:
            data["model"] = model
        if workspace_id:
            data["workspace_id"] = workspace_id

        return await self._client._request("POST", "/api/costs/constraints/check", json=data)

    # =========================================================================
    # Forecasting
    # =========================================================================

    async def get_forecast(
        self,
        workspace_id: str | None = None,
        days: int | None = None,
    ) -> dict[str, Any]:
        """Get basic cost forecast."""
        params: dict[str, Any] = {}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if days is not None:
            params["days"] = days

        return await self._client._request("GET", "/api/costs/forecast", params=params)

    async def get_detailed_forecast(
        self,
        workspace_id: str | None = None,
        days: int | None = None,
        include_alerts: bool = True,
    ) -> dict[str, Any]:
        """Get detailed forecast report with trend analysis."""
        params: dict[str, Any] = {"include_alerts": include_alerts}
        if workspace_id:
            params["workspace_id"] = workspace_id
        if days is not None:
            params["days"] = days

        return await self._client._request("GET", "/api/costs/forecast/detailed", params=params)

    async def simulate_scenario(
        self,
        workspace_id: str,
        scenario: dict[str, Any],
        days: int | None = None,
    ) -> dict[str, Any]:
        """Simulate a cost scenario."""
        data: dict[str, Any] = {
            "workspace_id": workspace_id,
            "scenario": scenario,
        }
        if days is not None:
            data["days"] = days

        return await self._client._request("POST", "/api/costs/forecast/simulate", json=data)
